From 3043b8423e1c0ea45cd3e4f4eb3d080757e18c26 Mon Sep 17 00:00:00 2001
From: ARATA Mizuki <minorinoki@gmail.com>
Date: Sun, 12 Jul 2015 21:51:35 +0900
Subject: [PATCH 2/5] Fix wxString::AsChar.

wxString::AsChar assumed that wxMBConv::FromWChar(NULL, 0, ...)
returns the real length of the converted string, which is not true
for wxMBConvUTF32* if wchar_t is UTF-16.
---
 src/common/string.cpp | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/src/common/string.cpp b/src/common/string.cpp
index 8b81d86..b8c8238 100644
--- a/src/common/string.cpp
+++ b/src/common/string.cpp
@@ -538,21 +538,25 @@ const char *wxString::AsChar(const wxMBConv& conv) const
     const size_t lenWC = m_impl.length();
 #endif // wxUSE_UNICODE_UTF8/wxUSE_UNICODE_WCHAR
 
-    const size_t lenMB = conv.FromWChar(NULL, 0, strWC, lenWC);
-    if ( lenMB == wxCONV_FAILED )
+    // The length of the needed buffer
+    const size_t bufLenMB = conv.FromWChar(NULL, 0, strWC, lenWC);
+    if ( bufLenMB == wxCONV_FAILED )
         return NULL;
 
-    if ( !m_convertedToChar.m_str || lenMB != m_convertedToChar.m_len )
+    if ( !m_convertedToChar.m_str || bufLenMB != m_convertedToChar.m_len )
     {
-        if ( !const_cast<wxString *>(this)->m_convertedToChar.Extend(lenMB) )
+        if ( !const_cast<wxString *>(this)->m_convertedToChar.Extend(bufLenMB) )
             return NULL;
     }
 
-    m_convertedToChar.m_str[lenMB] = '\0';
-    if ( conv.FromWChar(m_convertedToChar.m_str, lenMB,
-                        strWC, lenWC) == wxCONV_FAILED )
+    const size_t lenMB = conv.FromWChar(m_convertedToChar.m_str, bufLenMB,
+                                        strWC, lenWC);
+    if ( lenMB == wxCONV_FAILED )
         return NULL;
 
+    m_convertedToChar.m_str[lenMB] = '\0';
+    const_cast<wxString *>(this)->m_convertedToChar.m_len = lenMB;
+
     return m_convertedToChar.m_str;
 }
 
-- 
2.4.5

